cluster_management/src/main/java/com/pinterest/rocksplicator/utils/PartitionStateUpdater.java

```diff
@@ -0,0 +1,147 @@
+
+  class LeaderInfo {
+    String resource;
+    String partition;
+
+    @Override
+    public boolean equals(Object o) {        
+      if (this == o) return true; 
+      LeaderInfo li = (LeaderInfo) o;
+      return resource == li.resource && partition == li.partition;
+    }
+
+    @Override
+    public int hashCode() {
+      int hash = 0;
+      hash += resource == null ? 0 : resource.hashCode();
+      hash += partition == null ? 0 : partition.hashCode();
+      return hash;
+    }
+  } // End of LeaderInfo
+
+  public PartitionStateUpdater(
+    CuratorFramework zkClient,
+    String cluster,
+    int adminPort) {
+      this.zkClient = zkClient;
+      this.cluster = cluster;
+      this.adminPort = adminPort;
+  }
+  
+  public synchronized boolean shutdown() {
+    if (!running.get()) {
+      notify();
+      return false;
+    }
+    running.set(false);
+    notify();
+    return true;
+  }
+
+  public synchronized boolean addLeader(String resource, String partition) {
+    LeaderInfo li = new LeaderInfo();
+    li.partition = partition;
+    li.resource = resource;
+    if (leaders.containsKey(li)) {
+      return false;
+    }
+    leaders.put(li, new PartitionState());
+    notify();
+    return true;
+  }
+
+  public synchronized boolean removeLeader(String resource, String partition) {
+    LeaderInfo li = new LeaderInfo();
+    li.partition = partition;
+    li.resource = resource;
+    boolean success = leaders.remove(li) != null;
+    notify();
+    return success;
+  }
+
+  public void run() {
+    running.set(true);
+    LOG.error("PartitionStateUpdater started");
+    long waitTime = minWaitMs;
+    while(running.get()) {
+      synchronized(this) {
+        // update the states for all leaders
+        for (LeaderInfo leader : leaders.keySet()) {
+          try {
+            String dbName = Utils.getDbName(leader.partition);
+            long localSeq = Utils.getLocalLatestSequenceNumber(dbName, adminPort);
+            PartitionState lastState = leaders.get(leader);
+
+            // check if we have already updated a higher seqnum.
+            if (lastState.seqNum >= localSeq ) {
+              // we dont need to update
+              LOG.error(String.format("[%s] No need to update old:%d new:%d", leader.partition, lastState.seqNum, localSeq));
+              continue;
+            } else {
+              LOG.error(String.format("[%s] Updating old:%d new:%d", leader.partition, lastState.seqNum, localSeq));
+              lastState.setSeqNum(localSeq);
+              PartitionStateManager stateManager = new PartitionStateManager(cluster, leader.resource, leader.partition, zkClient);
+              stateManager.saveSeqNum(localSeq);
+            }
+          } catch(Exception e) {
+            LOG.error(String.format("Exception while storing state : %s",e));
+          }
+        }
+
+        // calculate wait time
+        runCount++;
+
+        if (runCount % 8 == 0) {
+          waitTime = minWaitMs;
```

Comment:
Reset runCount too?

Reply:
We could do that. But not needed as every 8 times, it will be in the same state. And we can use this later to see how many times this has been run

Reply:
should we reset `waitTime` for every 8 runs, or should we reset it for every leader add/removal? 
- if later, which might tell us, how often a leader info is changed